### ISSUE TITLE
tmux master has issues with character width on UTF-8 characters coupled with EOL wrap?  timing?

### ISSUE THREAD

**Issue by cgull** [tmux/tmux#539]
Current tmux master source at 895f1d93d51f54b3f76e6406d51f4fc491399ce2 seems to have problems with character width at the end of line.  In this example it occurs on UTF-8 codepoint U+00B0 (but I don't believe the specific codepoint matters).

Repro:
Start tmux, then:
```
git clone https://github.com/mobile-shell/mosh
cd mosh
git checkout mosh-1.2.6
src/tests/emulation-ascii-iso-8859.test baseline
```
Observed output:
```
20   21 ! 22 " 23 # 24 $ 25 % 26 & 27 ' 28 ( 29 ) 2a * 2b + 2c , 2d - 2e . 2f / 30 0 31 1 32 2 33 3 34 4 35 5 36 6 37 7 38 8 39 9 3a : 3b ; 3c < 3d = 3e > 3f ? 40 @ 41 A 42 B 43 C 44 D 45 E 46 F 47 G 48 H 49 I 4a J 4b K 4c L 4d M 4e N 4f O 50 P 51 Q 52 R 53 S 54 T 55 U 56 V 57 W 58 X 59 Y 5a Z 5b [ 5c \ 5d ] 5e ^ 5f _ 60 ` 61 a 62 b 63 c 64 d 65 e 66 f 67 g 68 h 69 i 6a j 6b k 6c l 6d m 6e n 6f o 70 p 71 q 72 r 73 s 74 t 75 u 76 v 77 w 78 x 79 y 7a z 7b { 7c | 7d } 7e ~ a0   a1 ¡ a2 ¢ a3 £ a4 ¤ a5 ¥ a6 ¦ a7 § a8 ¨ a9 © aa ª ab « ac ¬ ad ­ ae ® af ¯ b0 °b1 ± b2 ² b3 ³ b4 ´ b5 µ b6 ¶ b7 · b8 ¸ b9 ¹ ba º bb » bc ¼ bd ½ be ¾ bf ¿ c0 À c1 Á c2 Â c3 Ã c4 Ä c5 Å c6 Æ c7 Ç c8 È c9 É ca Ê cb Ë cc Ì cd Í ce Î cf Ï d0 Ð d1 Ñ d2 Ò d3 Ó d4 Ô d5 Õ d6 Ö d7 × d8 Ø d9 Ù da Ú db Û dc Ü dd Ý de Þ df ß e0 à e1 á e2 â e3 ã e4 ä e5 å e6 æ e7 ç e8 è e9 é ea ê eb ë ec ì ed í ee î ef ï f0 ð f1 ñ f2 ò f3 ó f4 ô f5 õ f6 ö f7 ÷ f8 ø f9 ù fa ú fb û fc ü fd ý fe þ ff ÿ
```
I discovered this while running Mosh's regression tests on OpenBSD 6.0 and this version of tmux (OpenBSD's bundled tmux is too old for Mosh's tests).  The problem is not consistently seen in the test scripting, so there may be some timing dependency or other influence on this problem.

Altering this test script to print a newline before printing characters whose low nibble is 0 (thus producing <80-char lines) clears or masks this problem.

I also see this when running the script by hand on FreeBSD 11.0-BETA4 and OS X 10.11.

My interactive terminals and Mosh's tests' tmux invocation both have 80-char-wide windows.  The exact location of corruption seems to vary with the width of the window.

**Comment by nicm** [comment]
When I run this in 80-column rxvt-unicode 9.21 on Linux with and without
tmux I see no difference.

**Comment by cgull** [comment]
I tried this on Ubuntu 16.04 and a Macintosh running XQuartz, using XTerm and rxvt from the rxvt-unicode package (there's about 6 different possible rxvt packages to install on Ubuntu).

With rxvt, I get similar results with and without rxvt (though both botch U+00AD-- that seems to be a separate issue from the one I'm seeing).  With xterm, I get a correct display standalone, but I see the end-of-line corruption with tmux.
![screen shot 2016-09-30 at 2 34 32 am](https://cloud.githubusercontent.com/assets/1934797/18982985/365902a6-86b8-11e6-9a60-074c7b61c43b.png)
![screen shot 2016-09-30 at 2 33 57 am](https://cloud.githubusercontent.com/assets/1934797/18982987/365cda0c-86b8-11e6-8ca7-4ac50a7e5787.png)
![screen shot 2016-09-30 at 2 33 06 am](https://cloud.githubusercontent.com/assets/1934797/18982984/3658f734-86b8-11e6-9942-279db2fbd91c.png)
![screen shot 2016-09-30 at 2 32 04 am](https://cloud.githubusercontent.com/assets/1934797/18982986/365bcf68-86b8-11e6-898c-62f865c3f33c.png)

**Comment by cgull** [comment]
That's with tmux built from bb7cc8e8fd9d797b9b00ab8254e51f25ca2db747.

**Comment by nicm** [comment]
A bunch of screenshots without context is not helpful, you need to be more specific about which ones are wrong and in what way.

**Comment by nicm** [comment]
I tried this in gnome-terminal and xterm and it looks fine there too for me.

**Comment by nicm** [comment]
Actually wait I might see a difference in gnome-terminal now.

**Comment by nicm** [comment]
The differences I see in gnome-terminal look like disagreement about the width of a soft hyphen (0xad).

tmux believes the system wcwidth(), which on Linux says width 0. It looks like rxvt-unicode does too, so tmux and rxvt-unicode match for me. gnome-terminal appears to think soft hyphen is width 1; if I change tmux to believe that too, then they match.

(I can't get xterm to behave sensibly without tmux, but I use it as my default terminal and I don't use UTF-8 so it may be weirdly configured, anyway I expect the problem is the same.)

If tmux and the terminal disagree about the width of a character, the cursor positioning will be wrong. You can see in your screenshots - in the last two (rxvt-unicode I assume), the hyphen is combined onto the previous space, so where the script expects two characters " -" tmux and rxvt-unicode are seeing one "-". But tmux and rxvt-unicode agree on the width, so they match.

In the first two screenshots (xterm), tmux assumes the terminal will output one character for " -" but in fact it outputs two, so tmux gets the cursor position wrong near the end of the line. In this case, the terminal is working the way the script itself expects, so the output looks "right" outside tmux.

**Comment by nicm** [comment]
I don't think this is fixable, there is no sensible way for us to predict that a terminal will treat a soft hyphen differently than we expect, and I don't see a way to make it the same across all terminals except by substituting (or ignoring) soft hyphen entirely.
